### ISSUE TITLE
RuntimeError

### ISSUE THREAD

**Issue by Katikonst** [javyduck/ChatScene#12]
RuntimeError: time-out of 10000ms while waiting for the simulator, make sure the simulator is ready and connected to 127.0.0.1:2000.
When running the examples of the modes, after a while it ends with this error.Also,I have open CARLA terminal and this error also occures for other scripts such as manual_control.py or generate_traffic.RAM and GPU seem okay and CARLA is running properly. Do you have any suggestions for this issue?

**Comment by javyduck** [comment]
hi, have you checked the port for your CARLA? is that running on the port 2000?

**Comment by Katikonst** [comment]
> hi, have you checked the port for your CARLA? is that running on the port 2000?

Hello,yeah. It is connected to port 2000

**Comment by javyduck** [comment]
hmmmm i see, you can first try the command `netstat -an | grep 2000` or `lsof -i :2000` to see if carla is indeed running on port 2000.

If it still does not work, then you can rerun the carla with another port like 2002, and specify the `--port 2002` when running the code to see if it will work.

**Comment by Katikonst** [comment]
> `netstat -an | grep 2000` or `lsof -i :2000`

i am actually having the same problem and it seems that it is connected to port 2002. I did the same for port 2000 as well. I appreciate your quick responses!

**Comment by javyduck** [comment]
sure, you specify the port (both carla and the parameter --port in code) to 2002 and try again.

**Comment by Katikonst** [comment]
> sure, you specify the port (both carla and the parameter --port in code) to 2002 and try again.

I have already specified and run it on port 2002. Unfortunately,the result is the same,even after our conversation from the above:RuntimeError: time-out of 10000ms while waiting for the simulator, make sure the simulator is ready and connected to 127.0.0.1:2002. Carla is connected.

**Comment by javyduck** [comment]
Could you kindly share the code for initiating the carla?

**Comment by Katikonst** [comment]
> Could you kindly share the code for initiating the carla?

Sure,it is the carla_runner.py file of ChatScene,right? Because I am following all the instructions to run carla with:# Launch CARLA
./CarlaUE4.sh -prefernvidia -windowed -carla-port=2000 or 2002 and train scenario with python scripts/run_train.py --agent_cfg=adv_scenic.yaml --scenario_cfg=train_scenario_scenic.yaml --mode train_scenario --scenario_id 1.

**Comment by javyduck** [comment]
yes, it's correct.

First, you should open a terminal to keep `./CarlaUE4.sh -prefernvidia -windowed -carla-port=2000` running, and you should not turn it off.

Then, open a **new** terminal to run `python scripts/run_train.py --agent_cfg=adv_scenic.yaml --scenario_cfg=train_scenario_scenic.yaml --mode train_scenario --scenario_id 1 --port 2000`.

**Comment by Katikonst** [comment]
> yes, it's correct.
> 
> First, you should open a terminal to keep `./CarlaUE4.sh -prefernvidia -windowed -carla-port=2000` running, and you should not turn it off.
> 
> Then, open a **new** terminal to run `python scripts/run_train.py --agent_cfg=adv_scenic.yaml --scenario_cfg=train_scenario_scenic.yaml --mode train_scenario --scenario_id 1 --port 2000`.

My apologies for not being specific enough. The problem is not from the start. Some scenes are being displayed and the message Beginning simulation of top-level scenario...
>> Loading scenario data id: 2
,alongside bev are working,but after a while it is showing the runtime error.

**Comment by javyduck** [comment]
it is kind of weird from my side... is your system ubuntu20.04? I would recommend to re-construct the whole env including CARLA and try again.

**Comment by Katikonst** [comment]
> it is kind of weird from my side... is your system ubuntu20.04? I would recommend to re-construct the whole env including CARLA and try again.

Yeah,i use ubuntu 20.04. I have already recontstructed the environment 2 times some weeks ago and I get the same result.Data are being saved,but the command of train_scenario and train_agent are not running all the way.

**Comment by javyduck** [comment]
oh, I see the problem, it is due to the inherent instability of CARLA instead of our repo.

Here is a workaround:

You can create a new python file (e.g., carla.py) under CARLA directory and put code:

```
import os
while True:
    os.system('./CarlaUE4.sh -prefernvidia -windowed -carla-port=2000')
```
And then using `python carla.py` under CARLA instead of directly running `./CarlaUE4.sh -prefernvidia -windowed -carla-port=2000`

You can just run it on the backend, and in this way, every time when the CARLA is broken down, it will automatically launch it again within the 10000ms waiting time.

**Comment by Katikonst** [comment]
> oh, I see the problem, it is due to the inherent instability of CARLA instead of our repo.
> 
> Here is a workaround:
> 
> You can create a new python file (e.g., carla.py) under CARLA directory and put code:
> 
> ```
> import os
> while True:
>     os.system('./CarlaUE4.sh -prefernvidia -windowed -carla-port=2000')
> ```
> 
> And then using `python carla.py` under CARLA instead of directly running `./CarlaUE4.sh -prefernvidia -windowed -carla-port=2000`
> 
> You can just run it on the backend, and in this way, every time when the CARLA is broken down, it will automatically launch it again within the 10000ms waiting time.

yeah, I was suspecting an instability in CARLA. I made a python file,but it is not reconnecting to the carla.py when it crashes. I tried to use nohup python carla.py & 
as well,but the train_scenario is still giving me the same error.Is there any other workaround?

**Comment by javyduck** [comment]
>  I made a python file,but it is not reconnecting to the carla.py when it crashes.

The carla.py file is for ensuring the carla keep running on the backend and will be relaunched whenever there is any error (just please make sure the carla is always running on the backend since sometimes the CARLA itself will crash).

For train_scenario, you can also do it in the same way:

create a new python file called `repeat_scenario.py`, and put:
```
import argparse
import os

def main():
    parser = argparse.ArgumentParser(description="Run training scenarios with and without resume.")
    # Setting default values as specified in your initial command
    parser.add_argument('--agent_cfg', type=str, default='adv_scenic.yaml', help='Configuration file for the agent.')
    parser.add_argument('--scenario_cfg', type=str, default='train_scenario_scenic.yaml', help='Configuration file for the scenario.')
    parser.add_argument('--mode', type=str, default='train_scenario', help='Mode of operation.')
    parser.add_argument('--scenario_id', type=int, default=1, help='Scenario identifier.')
    parser.add_argument('--port', type=int, default=2000, help='Port number for CARLA server.')
    parser.add_argument('--traffic_port', type=int, default=8000, help='Port number for traffic manager.')
    parser.add_argument('--repeat_times', type=int, default=5, help='Number of times to repeat resume runs.')

    args = parser.parse_args()

    # Initial run without resume
    os.system(f'python scripts/run_train.py --agent_cfg={args.agent_cfg} --scenario_cfg={args.scenario_cfg} '
              f'--mode={args.mode} --scenario_id={args.scenario_id} --port={args.port} --traffic_port={args.traffic_port}')

    # Repeated runs with resume if there is any crash
    for i in range(args.repeat_times):
        os.system(f'python scripts/run_train.py --agent_cfg={args.agent_cfg} --scenario_cfg={args.scenario_cfg} '
                  f'--mode={args.mode} --scenario_id={args.scenario_id} --port={args.port} --traffic_port={args.traffic_port} --resume')

if __name__ == "__main__":
    main()
```
Then use `python repeat_scenario.py --agent_cfg=adv_scenic.yaml --scenario_cfg=train_scenario_scenic.yaml --mode train_scenario --scenario_id 1` instead.

**Comment by Katikonst** [comment]
> hi, have you checked the port for your CARLA? is that running on the port 2000?

can I ask,do I need an OPENAI API KEY for retrieve.py? Is there a workaround?

**Comment by javyduck** [comment]
the default setting is directly doing retrieval from the database, so it does not need the openai key.

The openai key is used for generating some scenarios which can not be directly implemented using pure database, and you can also try to use other language models like llama3.

**Comment by Katikonst** [comment]
> the default setting is directly doing retrieval from the database, so it does not need the openai key.
> 
> The openai key is used for generating some scenarios which can not be directly implemented using pure database, and you can also try to use other language models like llama3.

which means that i could try llama3 to generate my own scenarios,if i don't have an openai key?

**Comment by javyduck** [comment]
yes, but i doubt the performance of llama3 would be extremely bad, and you may need to edit the generated scenario code file manually a bit to advoid some syntax error... (if it fails, then the generated file would be ended with 'txt' instead of 'scenic' for debugging)

**Comment by AHPUymhd** [comment]
> RuntimeError: time-out of 10000ms while waiting for the simulator, make sure the simulator is ready and connected to 127.0.0.1:2000. When running the examples of the modes, after a while it ends with this error.Also,I have open CARLA terminal and this error also occures for other scripts such as manual_control.py or generate_traffic.RAM and GPU seem okay and CARLA is running properly. Do you have any suggestions for this issue?

Hello, may I ask you a question? Have you solved this problem? I've been troubled by this issue for a long time. If you solve it, could you let me know how you did it? Thank you very much for your help.

**Comment by wardstoneX** [comment]
> > I made a python file,but it is not reconnecting to the carla.py when it crashes.
> 
> The carla.py file is for ensuring the carla keep running on the backend and will be relaunched whenever there is any error (just please make sure the carla is always running on the backend since sometimes the CARLA itself will crash).
> 
> For train_scenario, you can also do it in the same way:
> 
> create a new python file called `repeat_scenario.py`, and put:
> 
> ```
> import argparse
> import os
> 
> def main():
>     parser = argparse.ArgumentParser(description="Run training scenarios with and without resume.")
>     # Setting default values as specified in your initial command
>     parser.add_argument('--agent_cfg', type=str, default='adv_scenic.yaml', help='Configuration file for the agent.')
>     parser.add_argument('--scenario_cfg', type=str, default='train_scenario_scenic.yaml', help='Configuration file for the scenario.')
>     parser.add_argument('--mode', type=str, default='train_scenario', help='Mode of operation.')
>     parser.add_argument('--scenario_id', type=int, default=1, help='Scenario identifier.')
>     parser.add_argument('--port', type=int, default=2000, help='Port number for CARLA server.')
>     parser.add_argument('--traffic_port', type=int, default=8000, help='Port number for traffic manager.')
>     parser.add_argument('--repeat_times', type=int, default=5, help='Number of times to repeat resume runs.')
> 
>     args = parser.parse_args()
> 
>     # Initial run without resume
>     os.system(f'python scripts/run_train.py --agent_cfg={args.agent_cfg} --scenario_cfg={args.scenario_cfg} '
>               f'--mode={args.mode} --scenario_id={args.scenario_id} --port={args.port} --traffic_port={args.traffic_port}')
> 
>     # Repeated runs with resume if there is any crash
>     for i in range(args.repeat_times):
>         os.system(f'python scripts/run_train.py --agent_cfg={args.agent_cfg} --scenario_cfg={args.scenario_cfg} '
>                   f'--mode={args.mode} --scenario_id={args.scenario_id} --port={args.port} --traffic_port={args.traffic_port} --resume')
> 
> if __name__ == "__main__":
>     main()
> ```
> 
> Then use `python repeat_scenario.py --agent_cfg=adv_scenic.yaml --scenario_cfg=train_scenario_scenic.yaml --mode train_scenario --scenario_id 1` instead.

Hi, i looked at scenic_runner class and saw that continue_scenario_training wasnt being used, does it mean that we cant continiue scenario training if something crashes ?

**Comment by javyduck** [comment]
exactly, you are correct! "--cat" is used for resuming the training if there is any crash of CLARA.

**Comment by javyduck** [comment]
> > > I made a python file,but it is not reconnecting to the carla.py when it crashes.
> > 
> > 
> > The carla.py file is for ensuring the carla keep running on the backend and will be relaunched whenever there is any error (just please make sure the carla is always running on the backend since sometimes the CARLA itself will crash).
> > For train_scenario, you can also do it in the same way:
> > create a new python file called `repeat_scenario.py`, and put:
> > ```
> > import argparse
> > import os
> > 
> > def main():
> >     parser = argparse.ArgumentParser(description="Run training scenarios with and without resume.")
> >     # Setting default values as specified in your initial command
> >     parser.add_argument('--agent_cfg', type=str, default='adv_scenic.yaml', help='Configuration file for the agent.')
> >     parser.add_argument('--scenario_cfg', type=str, default='train_scenario_scenic.yaml', help='Configuration file for the scenario.')
> >     parser.add_argument('--mode', type=str, default='train_scenario', help='Mode of operation.')
> >     parser.add_argument('--scenario_id', type=int, default=1, help='Scenario identifier.')
> >     parser.add_argument('--port', type=int, default=2000, help='Port number for CARLA server.')
> >     parser.add_argument('--traffic_port', type=int, default=8000, help='Port number for traffic manager.')
> >     parser.add_argument('--repeat_times', type=int, default=5, help='Number of times to repeat resume runs.')
> > 
> >     args = parser.parse_args()
> > 
> >     # Initial run without resume
> >     os.system(f'python scripts/run_train.py --agent_cfg={args.agent_cfg} --scenario_cfg={args.scenario_cfg} '
> >               f'--mode={args.mode} --scenario_id={args.scenario_id} --port={args.port} --traffic_port={args.traffic_port}')
> > 
> >     # Repeated runs with resume if there is any crash
> >     for i in range(args.repeat_times):
> >         os.system(f'python scripts/run_train.py --agent_cfg={args.agent_cfg} --scenario_cfg={args.scenario_cfg} '
> >                   f'--mode={args.mode} --scenario_id={args.scenario_id} --port={args.port} --traffic_port={args.traffic_port} --resume')
> > 
> > if __name__ == "__main__":
> >     main()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then use `python repeat_scenario.py --agent_cfg=adv_scenic.yaml --scenario_cfg=train_scenario_scenic.yaml --mode train_scenario --scenario_id 1` instead.
> 
> Hi, i looked at scenic_runner class and saw that continue_scenario_training wasnt being used, does it mean that we cant continiue scenario training if something crashes ?

exactly, you are correct! "--cat" is used for resuming the training if there is any crash of CLARA.
